sdk/test/opentelemetry/sdk/trace/span_test.rb

```diff
@@ -7,4 +7,220 @@
+      span.finish
+      span.add_event(name: 'will_not_be_added')
+      span.to_proto[:events].must_be_nil
+    end
+  end
+
+  describe '#status=' do
+    it 'sets the status' do
+      span.status = Status.new(1, description: 'cancelled')
+      span.to_proto[:status].description.must_equal('cancelled')
+    end
+
+    it 'does not set the status if span is ended' do
+      span.finish
+      span.status = Status.new(1, description: 'cancelled')
+      span.to_proto[:status].must_be_nil
+    end
+  end
+
+  describe '#name=' do
+    it 'sets the name' do
+      span.name = 'new_name'
+      span.to_proto[:name].must_equal('new_name')
+    end
+
+    it 'does not set the name if span is ended' do
+      span.finish
+      span.name = 'new_name'
+      span.to_proto[:name].must_equal('name')
+    end
+  end
+
+  describe '#finish' do
+    it 'returns itself' do
+      span.finish.must_equal(span)
+    end
+
+    it 'sets the end timestamp' do
+      span.finish
+      span.to_proto[:end_timestamp].wont_be_nil
+    end
+
+    it 'calls the span processor #on_end callback' do
+      mock_span_processor.expect(:on_start, nil) do |_|
+        true
+      end
+      span = Span.new(context, 'name', SpanKind::INTERNAL, nil, trace_config,
+                      mock_span_processor, nil, nil, nil, Time.now)
+      mock_span_processor.expect :on_end, nil, [span]
+      span.finish
+      mock_span_processor.verify
+    end
+
+    it 'marks the span as ended' do
+      span.finish
+      span.instance_variable_get(:@ended).must_equal(true)
+    end
+
+    it 'does not allow ending more than once' do
+      span.finish
+      span.instance_variable_get(:@ended).must_equal(true)
+      ts = span.to_proto[:end_timestamp]
+      span.finish
+      span.to_proto[:end_timestamp].must_equal(ts)
+    end
+  end
+
+  describe '#initialize' do
+    it 'installs the span processor' do
+      span.instance_variable_get(:@span_processor).must_equal(span_processor)
+    end
+
+    it 'calls the span processor #on_start callback' do
+      mock_span_processor.expect(:on_start, nil) do |_|
+        true
+      end
+      Span.new(context, 'name', SpanKind::INTERNAL, nil, trace_config,
+               mock_span_processor, nil, nil, nil, Time.now)
+      mock_span_processor.verify
+    end
+
+    it 'trims excess attributes' do
+      attributes = { 'foo': 'bar', 'other': 'attr' }
+      span = Span.new(context, 'name', SpanKind::INTERNAL, nil, trace_config,
+                      span_processor, attributes, nil, nil, Time.now)
+      span.to_proto[:total_recorded_attributes].must_equal(2)
+      span.to_proto[:attributes].length.must_equal(1)
+    end
+
+    it 'counts attributes' do
+      attributes = { 'foo': 'bar', 'other': 'attr' }
+      span = Span.new(context, 'name', SpanKind::INTERNAL, nil, trace_config,
+                      span_processor, attributes, nil, nil, Time.now)
+      span.instance_variable_get(:@total_recorded_attributes).must_equal(2)
```

Comment:
`to_proto` will freeze attributes so I left this as an `instance_variable_get` rather than create the span twice in the test to test both ways attributes can be added